### PR TITLE
support build docker images for arm64 arch

### LINKED ISSUE #1851: Upload container images for arm64 architecture
Now that AWS announced their way more powerful M6g instances it will start to get interesting to move from x86 to the ARM architecture. You already provide binaries on the releases page, but the docker images are for x86 only. Could you maybe adjust your build pipeline and provide images for ARM like [prometheus does](https://hub.docker.com/r/prom/prometheus/tags)?

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Xiang Dai <764524258@qq.com>

* [x] I added CHANGELOG entry for this change.

## Changes

Fix #1851

### CODE REVIEW

**Comment by daixiang0** [comment]
I have no permission to test push related things, build pass at local.

**Comment by daixiang0** [comment]
Prometheus also support arm, but i saw corss build only generate binary for amd64 and arm64, so do not add it.

File: Makefile
Comment (bwplotka) [inline Makefile]: Why push twice?

Comment (daixiang0) [reply Makefile]: dockerhub and quay.io, notice that there are 2 images.

**Comment by daixiang0** [comment]
@bwplotka I use the way Prometheus does, please review again.

**Comment by simonpasquier** [comment]
A few things from my Prometheus experience:
* someone with permissions needs to create in advance the following repositories on quay.io.
  * quay.io/thanos/thanos-linux-amd64
  * quay.io/thanos/thanos-linux-arm64
* you need to add a step running `docker run --privileged linuxkit/binfmt:v0.7` (see [here](https://github.com/prometheus/circleci/blob/c9c999b0878f63830efd13d6180d6a54a4900859/src/commands/setup_build_environment.yml#L8)).
* you might need to upgrade the Docker version to `18.06.0-ce`.

And there might other things that I don't remember ;-)

**Comment by daixiang0** [comment]
@simonpasquier yep, but i have no premissions to test while process:(

File: .circleci/config.yml
Comment (daixiang0) [reply .circleci/config.yml]: It register qemu to support mulit-arch at amd arch.
I will do this.

**Comment by daixiang0** [comment]
@bwplotka wait for your feedback~

**Comment by daixiang0** [comment]
@bwplotka do you have time to test it?

**Comment by simonpasquier** [comment]
This looks good to me. The only real way to be sure it works would be to merge and see what happens :-)

**Comment by daixiang0** [comment]
@bwplotka mind have a look?

**Comment by daixiang0** [comment]
any input for this?

**Comment by daixiang0** [comment]
@bwplotka hi, do you have time to test it?

**Comment by bwplotka** [comment]
Yea not really.  Will speak offline how we can move this forward (:

**Comment by povilasv** [comment]
Hey, I'm interested in this and I think I will be able to help, ping me on monday :)

**Comment by daixiang0** [comment]
@povilasv thanks!

File: Dockerfile
Comment (povilasv) [inline Dockerfile]:
```suggestion
COPY .build/${OS}-${ARCH}/thanos /bin/thanos
```

File: Dockerfile
Comment (povilasv) [inline Dockerfile]: Let's add `ARG OS="linux"` here

**Comment by daixiang0** [comment]
The [CI test result](https://github.com/thanos-io/thanos/pull/1906/checks?check_run_id=651998085) is confused:
```
>> building binaries /go/bin
promu-9583e5a6448f97c6294dca72dd1d173e28f8d4a4: error: unknown long flag '--prefix', try --help
##[error]make: *** [Makefile:153: build] Error 1
```

File: CHANGELOG.md
Comment (povilasv) [inline CHANGELOG.md:175]: Let's remove this

File: Makefile
Comment (povilasv) [inline Makefile:154]: This is a bit confusing. I guess some other tools depend on `thanos` binary being in current directory?

Comment (daixiang0) [reply Makefile:154]: Yes, `scripts/genflagdocs` need it.

**Comment by daixiang0** [comment]
I think no need to create it for quay, details can refer to `publish-master` in ".circleci/config.yaml".

For dockerfhub, i am not sure since it is forked.

I have no access to run `publish-master` so need you help @povilasv .

**Comment by simonpasquier** [comment]
> We will need to create empty repos in quay/dockerhub:

You shouldn't need to create the arch-specific repositories in DockerHub. It should only be done for Quay. At least this is what we're doing for the prometheus organization.

File: .circleci/config.yml
Comment (simonpasquier) [inline .circleci/config.yml:76]: I'm not sure but it might be required to push the individual images before pushing the manifest. At least this is how we're doing it in Prometheus.

File: .circleci/config.yml
Comment (simonpasquier) [inline .circleci/config.yml:74]: I suspect it won't work because Circle CI doesn't cross-build binaries when merging on `master`.

**Comment by daixiang0** [comment]
still need it, would do this.

**Comment by cryptk** [comment]
Is there anything that can be done to move this along? I also have a need for arm64 thanos and I would love to help test it out.

**Comment by daixiang0** [comment]
still in process.

**Comment by daixiang0** [comment]
ping @bwplotka

**Comment by povilasv** [comment]
@daixiang0  can you  fix the conflicts ;)

**Comment by daixiang0** [comment]
@povilasv done

**Comment by daixiang0** [comment]
ping @povilasv @bwplotka

**Review by povilasv — APPROVED** [review]

**Comment by daixiang0** [comment]
@bwplotka could we merge this?

**Comment by daixiang0** [comment]
@bwplotka any chance to merge it?

**Comment by daixiang0** [comment]
ping @thanos-io/thanos-maintainers

**Comment by daixiang0** [comment]
@bwplotka any chance to merge it?

File: Makefile
Comment (bwplotka) [inline Makefile:9]: This is some Makefile magic I never seen before, do you mind explaining what it does? ;p preferably in comment?

**Review by bwplotka — APPROVED** [review]
I think there are many things which will break if we merge this (especially on upload path -> I doubt dockerhub supports this), but we will never know until we merge. I would be happy to merge this if you will help us @daixiang0 to fix issues immediately after this. Also one comment beforing merging would be nice :hugs: 

It's true, it's worth to support ARM containers, why not.

File: Makefile
Comment (daixiang0) [reply Makefile:9]: I will help fix :)

Add them.

**Comment by bwplotka** [comment]
I think we merged already one PR, no?

**Comment by daixiang0** [comment]
No, it is for pinned image tag.

**Comment by foshantiger** [comment]
Is this PR will be merged？

**Comment by DerrickMartinez** [comment]
Any progress on this?

**Comment by commixon** [comment]
Hey people, any progress on this one?

**Comment by oded-dd** [comment]
Hey, It is blocking our move, Any update?

**Comment by daixiang0** [comment]
I would update soon

**Comment by GiedriusS** [comment]
@daixiang0 any update? Maybe someone else can pick this up :thinking:

**Comment by daixiang0** [comment]
@GiedriusS sorry for slow process, seems only CI issues left, I would do it today, hopefully finish too.

File: Makefile
Comment (onprem) [inline Makefile:194]: Don't we need to modify `$(DOCKER_IMAGE_REPO)` int these two lines with `$(DOCKER_IMAGE_REPO)-linux-$*` as well?

**Review by onprem — COMMENTED** [review]
This is looking good overall, thanks for doing this. I have a few comments.

File: Dockerfile.multi-arch
Comment (onprem) [inline Dockerfile.multi-arch:3]: I don't think using `latest` here is a good idea. Maybe we should pin this to a SHA as we are doing in the primary `Dockerfile`

File: Dockerfile.multi-arch
Comment (squat) [inline Dockerfile.multi-arch:1]: why do we need a new Dockerfile vs adding this functionality to the existing one? It seems a bit like a workaround right now that will cause confusion in the future.

Comment (ChipWolf) [reply Dockerfile.multi-arch:1]: Agree 100%

Comment (daixiang0) [reply Dockerfile.multi-arch:1]: Done.

File: Dockerfile.multi-arch
Comment (daixiang0) [reply Dockerfile.multi-arch:3]: Done.

**Comment by yurrriq** [comment]
Any update? I guess we'll plan to use https://hub.docker.com/r/raspbernetes/thanos for now, as mentioned on the other ticket.

**Comment by daixiang0** [comment]
@thanos-io/thanos-maintainers please review, thanks!

**Comment by GiedriusS** [comment]
Let's do this for the next release, I will take a look in a bit.

File: Makefile
Comment (GiedriusS) [inline Makefile:16]: The hashes here have been switched accidentally, let me fix ^

File: .circleci/config.yml
Comment (GiedriusS) [inline .circleci/config.yml:84]: Another thing: `docker manifest push ...` needs to happen _after_ pushing the Docker containers to a registry:
https://github.com/prometheus/circleci/blob/master/src/commands/publish_images.yml#L47-L48
https://github.com/docker/cli/issues/2396
https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

File: .circleci/config.yml
Comment (GiedriusS) [inline .circleci/config.yml:77]: Also you've forgotten to do the same for the `publish_release` job, let me do that

**Review by GiedriusS — CHANGES_REQUESTED** [review]
Actually, this still misses one crucial part - the same `thanos` binary is copied into both images. The `arm64` one needs to have the `arm64` binary :/ Will come back soon once I'll have some time :/

**Comment by yurrriq** [comment]
Any update?

**Comment by avivgold098** [comment]
Bump :)
This will be very useful

**Comment by yeya24** [comment]
Hello @GiedriusS, do you happen to have a chance to look at it recently?

**Comment by GiedriusS** [comment]
Unfortunately, but no. Help welcome.

**Comment by daixiang0** [comment]
@GiedriusS @hanjm Please review again.

File: Makefile
Comment (GiedriusS) [inline Makefile:180]: I think the problem is the same: here we are copying the same `thanos` binary into both images. We need to have the OS/ARCH as parameters https://github.com/prometheus/prometheus/blob/main/Dockerfile#L8 and use them when copying the binary. So, it should be something like:
```
make crossbuild # To build all binaries
```

And then `Dockerfile` needs to be like that example (:

File: CHANGELOG.md
Comment (GiedriusS) [inline CHANGELOG.md:404]: Rebase went wrong? (:

Comment (daixiang0) [reply CHANGELOG.md:404]: Sorry for mistake.

File: Makefile
Comment (daixiang0) [reply Makefile:180]: In fact Prometheus only build Linux OS:
https://github.com/prometheus/prometheus/blob/main/Makefile.common#L249-L256

Comment (daixiang0) [reply Makefile:180]: Now updated.

File: .circleci/config.yml
Comment (GiedriusS) [inline .circleci/config.yml:82]: Last improvement that I'd like to see before merging: this builds the binary for all supported archs/OSes but for `main` we only care about two - arm64/amd64. Could we only build those? It takes quite a long time to build for all OSes on my laptop so I can only imagine what it will be like on CircleCI :smile:

File: .dockerignore
Comment (GiedriusS) [inline .dockerignore:17]: Not sure how you built/tested the images without removing this (:

File: .circleci/config.yml
Comment (daixiang0) [reply .circleci/config.yml:82]: I have added the following logic: if on the main branch, only build above two.

File: Makefile
Comment (GiedriusS) [inline Makefile:146]: I had to do:
```
GIT_BRANCH=$(shell $(GIT) rev-parse --abbrev-ref HEAD)
```

Otherwise it wouldn't have a proper value (:

Comment (GiedriusS) [reply Makefile:146]: Added this fix!

**Review by GiedriusS — APPROVED** [review]
:muscle: hope this works

**Pull request merged**
